actors/ppo_atari_pz.py

```diff
@@ -0,0 +1,1085 @@
+    """Layer initialization"""
+    torch.nn.init.orthogonal_(layer.weight, std)
+    torch.nn.init.constant_(layer.bias, bias_const)
+    return layer
+
+
+class PolicyValueNetwork(torch.nn.Module):
+    """Policy and Value networks for Atari games"""
+
+    def __init__(self, num_actions: int) -> None:
+        super().__init__()
+        self.shared_network = torch.nn.Sequential(
+            initialize_layer(torch.nn.Conv2d(6, 32, 8, stride=4)),
+            torch.nn.ReLU(),
+            initialize_layer(torch.nn.Conv2d(32, 64, 3, stride=2)),
+            torch.nn.ReLU(),
+            initialize_layer(torch.nn.Conv2d(64, 64, 3, stride=1)),
+            torch.nn.ReLU(),
+            torch.nn.Flatten(),
+            initialize_layer(torch.nn.Linear(64 * 7 * 7, 512)),
+            torch.nn.ReLU(),
+        )
+        self.actor = initialize_layer(torch.nn.Linear(512, num_actions), 0.01)
+        self.value = initialize_layer(torch.nn.Linear(512, 1), 1)
+
+    def get_value(self, observation: torch.Tensor) -> torch.Tensor:
+        """Compute the value of being in a state"""
+        observation = observation / 255.0
+        return self.value(self.shared_network(observation))
+
+    def get_action(self, observation: torch.Tensor) -> Distribution:
+        """Actions given observations"""
+        observation = observation / 255.0
+        action_logits = self.actor(self.shared_network(observation))
+        dist = torch.distributions.categorical.Categorical(logits=action_logits)
+
+        return dist
+
+
+class PPOModel(Model):
+    """Proximal Policy Optimization (PPO) is an on-policy algorithm.
+    https://arxiv.org/pdf/1707.06347.pdf.
+
+    Attributes:
+
+    """
+
+    def __init__(
+        self,
+        model_id: int,
+        environment_implementation: str,
+        num_actions: int,
+        input_shape: tuple,
+        num_policy_outputs: int,
+        learning_rate: float = 0.01,
+        n_iter: int = 1000,
+        dtype=torch.float32,
+        device: str = "cpu",
+        version_number: int = 0,
+    ) -> None:
+
+        super().__init__(model_id, version_number)
+        self._environment_implementation = environment_implementation
+        self._num_actions = num_actions
+        self.input_shape = input_shape
+        self.num_policy_outputs = num_policy_outputs
+        self.device = device
+        self._dtype = dtype
+        self._n_iter = n_iter
+        self.network = PolicyValueNetwork(num_actions=self._num_actions).to(self._dtype)
+        self.network.to(torch.device(self.device))
+
+        # Get optimizer for two models
+        self.network_optimizer = torch.optim.Adam(self.network.parameters(), lr=learning_rate, eps=1e-5)
+        # lambda1 = lambda epoch: 0.98 ** epoch
+        # self.scheduler = torch.optim.lr_scheduler.LambdaLR(self.network_optimizer, lr_lambda=lambda1)
+        self.scheduler = torch.optim.lr_scheduler.StepLR(self.network_optimizer, step_size=1000, gamma=0.1)
```

Comment:
I'm not sure what the point of storing that in the model is, especially as they are not saved. They'd fit more naturally in the training loop no?

Reply:
`self.network_optimizer` and `self.scheduler` were moved to `__init__` constructor in the training loop